File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/PaymentStrategy/CryptoCurrencyStrategy.cs
using Assignment2Solution.Domain.Inventory;

namespace Assignment2Solution.Domain.PaymentStrategy;

/// <summary>
///     A payment strategy for cryptocurrency payments.
/// </summary>
public sealed class CryptoCurrencyStrategy : PaymentStrategyBase
{
    private readonly string _transactionSignature;
    private readonly string _walletAddress;

    /// <summary>
    ///     Initializes a new instance of the <see cref="CryptoCurrencyStrategy" /> class.
    /// </summary>
    /// <param name="walletAddress">The cryptocurrency wallet address.</param>
    /// <param name="transactionSignature">The transaction signature.</param>
    /// <exception cref="ArgumentException">Thrown when wallet address or transaction signature is empty.</exception>
    public CryptoCurrencyStrategy(string walletAddress, string transactionSignature)
    {
        if (string.IsNullOrWhiteSpace(walletAddress))
            throw new ArgumentException("Wallet address cannot be empty.", nameof(walletAddress));
        if (string.IsNullOrWhiteSpace(transactionSignature))
            throw new ArgumentException("Transaction signature cannot be empty.", nameof(transactionSignature));

        _walletAddress = walletAddress;
        _transactionSignature = transactionSignature;
    }

    /// <inheritdoc />
    public override void Checkout(InventoryItem item, int quantity, TextWriter output)
    {
        var total = ComputeTotalAmount(item, quantity);
        output.WriteLine(
            $"Checking out {total:C} using CryptoCurrency (Wallet: {_walletAddress}, Signature: {_transactionSignature}).");
    }
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/payment/CryptoCurrencyStrategy.java
package assignment2solution.domain.payment;

import assignment2solution.domain.inventory.InventoryItem;

import java.io.PrintWriter;
import java.io.Writer;
import java.text.NumberFormat;
import java.util.Locale;
import java.util.Objects;

/**
 * A payment strategy for cryptocurrency payments.
 */
public final class CryptoCurrencyStrategy extends PaymentStrategyBase {
    private final String walletAddress;
    private final String transactionSignature;
    private final NumberFormat currencyFormatter = NumberFormat.getCurrencyInstance(Locale.US);

    /**
     * Initializes a new instance of the {@link CryptoCurrencyStrategy} class.
     */
    public CryptoCurrencyStrategy(String walletAddress, String transactionSignature) {
        if (walletAddress == null || walletAddress.trim().isEmpty()) {
            throw new IllegalArgumentException("Wallet address cannot be empty.");
        }
        if (transactionSignature == null || transactionSignature.trim().isEmpty()) {
            throw new IllegalArgumentException("Transaction signature cannot be empty.");
        }
        this.walletAddress = walletAddress;
        this.transactionSignature = transactionSignature;
    }

    @Override
    public void checkout(InventoryItem item, int quantity, Writer output) {
        Objects.requireNonNull(item, "item");
        Objects.requireNonNull(output, "output");
        var writer = output instanceof PrintWriter ? (PrintWriter) output : new PrintWriter(output, true);
        var total = computeTotalAmount(item, quantity);
        writer.println("Checking out " + currencyFormatter.format(total)
            + " using CryptoCurrency (Wallet: " + walletAddress + ", Signature: " + transactionSignature + ").");
    }
}
